static-check/src/types/type_.rs

```diff
@@ -0,0 +1,330 @@
+            _ => {}
+        }
+    }
+}
+
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self {
+            Type::Address => write!(f, "address"),
+            Type::Boolean => write!(f, "bool"),
+            Type::Field => write!(f, "field"),
+            Type::Group => write!(f, "group"),
+            Type::IntegerType(integer_type) => write!(f, "{}", integer_type),
+
+            Type::Array(type_, dimensions) => {
+                let dimensions_string = dimensions
+                    .iter()
+                    .map(|dimension| format!("{}", dimension))
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                write!(f, "[{}; ({})]", *type_, dimensions_string)
+            }
+            Type::Tuple(tuple) => {
+                let tuple_string = tuple.iter().map(|x| format!("{}", x)).collect::<Vec<_>>().join(", ");
+
+                write!(f, "({})", tuple_string)
+            }
+
+            Type::Circuit(identifier) => write!(f, "circuit {}", identifier),
+            Type::Function(identifier) => write!(f, "function {}", identifier),
+            Type::TypeVariable(type_variable) => write!(f, "{}", type_variable),
+        }
+    }
+}
+
+impl PartialEq for Type {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Type::Address, Type::Address) => true,
+            (Type::Boolean, Type::Boolean) => true,
+            (Type::Field, Type::Field) => true,
+            (Type::Group, Type::Group) => true,
+            (Type::IntegerType(integer_type1), Type::IntegerType(integer_type2)) => integer_type1.eq(integer_type2),
+
+            (Type::Array(type1, dimensions1), Type::Array(type2, dimensions2)) => {
+                // Flatten both array types before comparison.
+                let (type1_flat, dimensions1_flat) = flatten_array_type(type1, dimensions1.to_owned());
+                let (type2_flat, dimensions2_flat) = flatten_array_type(type2, dimensions2.to_owned());
+
+                // Element types and dimensions must match
+                type1_flat.eq(type2_flat) && dimensions1_flat.eq(&dimensions2_flat)
+            }
+
+            (Type::Tuple(types1), Type::Tuple(types2)) => types1.eq(types2),
+            (Type::Circuit(identifier1), Type::Circuit(identifier2)) => identifier1.eq(identifier2),
+            (Type::Function(identifier1), Type::Function(identifier2)) => identifier1.eq(identifier2),
+            (Type::TypeVariable(variable1), Type::TypeVariable(variable2)) => variable1.eq(variable2),
+            _ => false,
+        }
+    }
+}
+
+impl Eq for Type {}
+
+///
+/// Returns the data type of the array element and vector of dimensions.
+///
+/// Will flatten an array type `[[[u8; 1]; 2]; 3]` into `[u8; (3, 2, 1)]`.
+///
+pub fn flatten_array_type(type_: &Type, mut dimensions: Vec<usize>) -> (&Type, Vec<usize>) {
```

Comment:
```suggestion
pub fn flatten_array_type<'a>(type_: &Type, dimensions: Cow<'a, [usize]>) -> (&Type, Cow<'a, [usize]>) {
```

Reply:
I expect `dimensions` to often not require an allocation at all; a `Cow` is way cheaper than an allocation